contracts/external/policies/ValueLimitPolicy.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.23;
+
+import "contracts/interfaces/IPolicy.sol";
+import { CallType, CALLTYPE_BATCH, CALLTYPE_SINGLE, CALLTYPE_DELEGATECALL } from "erc7579/lib/ModeLib.sol";
+import { ISmartSession } from "contracts/ISmartSession.sol";
+import { IERC7579Account, Execution } from "erc7579/interfaces/IERC7579Account.sol";
+import { ExecutionLib } from "contracts/lib/ExecutionLib.sol";
+
+/**
+ * @title ValueLimitPolicy
+ * @notice A policy that limits the total amount of value that can be transferred within a permission.
+ * @dev This is Action Policy as well as UserOp Policy. So the separate limit can be set for each action
+ * in a permission or for all userOps in a permission.
+ */
+contract ValueLimitPolicy is IActionPolicy, IUserOpPolicy {
+
+    using ExecutionLib for bytes;
+
+    struct ValueLimitConfig {
+        uint256 valueLimit;
+        uint256 limitUsed;
+    }
+
+    mapping(ConfigId id => mapping(address msgSender => mapping(address userOpSender => ValueLimitConfig))) public
+        valueLimitConfigs;
+
+    /**
+     * @notice Checks if the value limit is not exceeded by a given user operation.
+     * @param id The config ID.
+     * @param op The user operation.
+     * @return The validation result.
+     */
+    function checkUserOpPolicy(ConfigId id, PackedUserOperation calldata op) external returns (uint256) {
+        ValueLimitConfig storage $config = valueLimitConfigs[id][msg.sender][op.sender];
+        require($config.valueLimit != 0, PolicyNotInitialized(id, msg.sender, op.sender));
+        bytes4 selector = bytes4(op.callData[0:4]);
+        if (selector == IERC7579Account.execute.selector) {
+            (CallType callType, ) = op.callData.get7579ExecutionTypes();
+            if (callType == CALLTYPE_SINGLE) {
+                (, uint256 value, ) = op.callData.decodeUserOpCallData().decodeSingle();
+                if (!_checkAndAdjustLimit($config, value)) return VALIDATION_FAILED;
+            } else if (callType == CALLTYPE_BATCH) {
+                Execution[] calldata executions = op.callData.decodeUserOpCallData().decodeBatch();
+                uint256 length = executions.length;
+                uint256 totalValue = 0;
+                for (uint256 i; i < length; i++) {
+                    totalValue += executions[i].value;
+                }
+                if (!_checkAndAdjustLimit($config, totalValue)) return VALIDATION_FAILED;
+            } else if (callType == CALLTYPE_DELEGATECALL) {
+                // delegatecall does not transfer value
+                return VALIDATION_SUCCESS;
+            } else {
+                // it doesn't return validation failed to express that validation was not possible
+                revert ISmartSession.UnsupportedCallType(callType);
```

Comment:
4337 says that validation shouldnt revert but just return false on errors

Reply:
https://eips.ethereum.org/EIPS/eip-4337#account-contract-interface

> If the account does not support signature aggregation, it MUST validate that the signature is a valid signature of the userOpHash, and SHOULD return SIG_VALIDATION_FAILED (and not revert) on signature mismatch. Any other error MUST revert.

I think that's exactly the 'revert' case here